README.md
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](/License.md)

language-bve5
===
BveTrainsim5.7用構文パッケージ in Atom

![syntaxhighlighting](https://raw.githubusercontent.com/aoisupersix/Atom_Bve5Package/master/images/syntax.png)

## これは何？

AtomでBveTrainsim5.7構文のシンタックスハイライト、スニペット挿入、その他制作に役立つ機能をサポートするパッケージです。  
BveTrainsim5.7以前の構文にも一部対応しています。

## インストール

Atomの設定画面から```language-bve5```と検索してインストールするか、```apm install language-bve5```を実行してください。

## シンタックスハイライト対応ファイル一覧

以下のファイル形式に対応しています。（括弧内は各ファイルのヘッダ)

- マップ(BveTs Map 2.02)
- ストラクチャーリスト(BveTs Structure List 2.00)
- サウンドリスト(BveTs  Sound List 2.00)
- 停車場リスト(BveTs Station List 2.00)
- 他列車(BveTs Train 1.01)

各構文に関しては、[BveTrainsim公式サイト](http://bvets.net)で確認してください。

## 文法の有効化

文法を有効化することによってパッケージの機能が利用できるようになります。  
文法の判定は、読み込み/編集時に各ファイルのヘッダから自動で行われますが、コマンドからも実行できます。

![grammarJudgment](https://raw.githubusercontent.com/aoisupersix/Atom_Bve5Package/master/images/grammarJudgment.gif)

文法の判定を行うには、コマンドパレットの```Language Bve:GrammarJudgment```を選択するか、メニューバーの```Packages```→```Language Bve```→```Judge Grammar```を選択してコマンドを実行してください。

自動で文法が判定されない場合、右下の```Plain Text```をクリックして、手動で文法を指定してください。

## スニペット挿入の利用

スニペット挿入とは繰り返し入力する記述を自動で入力してくれる機能です。  
BveTrainsim5の構文を途中まで入力するとスニペットの候補が出てくるので、```TAB```を押すことで構文を自動で入力、引数の位置へとカーソルを移動してくれます。

![snippet](https://raw.githubusercontent.com/aoisupersix/Atom_Bve5Package/master/images/snippet.gif)

このパッケージではマップファイルのほぼ全ての構文のスニペット挿入をサポートしています。

## 駅位置へジャンプ

マップファイル上で```Language Bve:JumpStation```コマンドを実行することにより、マップファイル内の指定した駅位置にカーソルを移動させることが出来ます。

![jumpStation](https://raw.githubusercontent.com/aoisupersix/Atom_Bve5Package/master/images/jumpStation.gif)

### ライセンス

>The MIT Lisense (MIT)
>
>Copyright (c) 2016 aoisupersix

* **ライセンス全文:** [LICENSE.md](https://github.com/aoisupersix/Atom_Bve5Package/blob/master/LICENSE.md)

## サポート

バグの報告や要望はIssue、もしくはこちらへ。
<http://aoisupersix.tokyo/>


CHANGELOG.md
## 0.1.0 - First Release
  Set Syntax highlighting.
  Set auto-complete.(Only a part)
## 0.1.1 - Bug Fix
  Fix Repeater construction
## 0.1.2 - Bug Fix
  Fix Package to 'bve5-map' from 'bve5-map2'
  Clean Directory
## 0.2.0 - Add new construction
  Add syntax grammar and snippets
  Add README.md
## 0.2.1 - Bug Fix
  Fix snippets code
## 1.0.0 - Add Support files & Fine improvement
  Add new support files
  Add automatic judgment of grammar
  Change package name and README.md
## 1.0.1 - Bug Fix
  Add missing files
## 1.1.0 - Fine improvement
  Improvement header automatic judgment
  Update header regular expression
  Change package name to 'language-bve5' from 'bve5-Syntax_Snippets'
## 1.2.0 - Add new function
  Add station name jump function
## 1.2.1 - Improve performance
  Change list view instantiation to lazy load
  Add Menu Items
## 1.2.2 - Bug Fix
  Correnction of wrong sentences in Readme.md
## 1.2.3 - Fine improvement
  Change images in Readme.md
  Update snippets
## 1.3.0 - Improve grammar and add snippet
  Improve accuracy of syntax highlighting
  Add all syntax snippets
## 1.3.1 - Detailed processing change
  Fixed a bug with different snippet characters
  Improve grammar automatic judgment processing
## 1.3.2 - Bug Fixed
  Fixed snippet bug.


spec/language-bve-map-spec.js
'use babel';

import LanguageBveMap from '../lib/language-bve-map';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('LanguageBveMap', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('language-bve-map');
  });

  describe('when the language-bve-map:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.language-bve-map')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'language-bve-map:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.language-bve-map')).toExist();

        let languageBveMapElement = workspaceElement.querySelector('.language-bve-map');
        expect(languageBveMapElement).toExist();

        let languageBveMapPanel = atom.workspace.panelForItem(languageBveMapElement);
        expect(languageBveMapPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'language-bve-map:toggle');
        expect(languageBveMapPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.language-bve-map')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'language-bve-map:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let languageBveMapElement = workspaceElement.querySelector('.language-bve-map');
        expect(languageBveMapElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'language-bve-map:toggle');
        expect(languageBveMapElement).not.toBeVisible();
      });
    });
  });
});


spec/language-bve-map-view-spec.js
'use babel';

import LanguageBveMapView from '../lib/language-bve-map-view';

describe('LanguageBveMapView', () => {
  it('has one valid test', () => {
    expect('life').toBe('easy');
  });
});


lib/language-bve.js
'use babel';

import { CompositeDisposable } from 'atom';
SearchStaView = null;

export default {

  staSearchListView: null,  //駅名検索ビュー
  subscriptions: null,

  activate(state) {

    // Events subscribed to in atom's system can be easily cleaned up with a CompositeDisposable
    this.subscriptions = new CompositeDisposable();

    // コマンド登録
    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'language-bve:GrammarJudgment': () => this.grammarJudgment(),
      'language-bve:JumpStation': () => this.openSearchStaPanel()
    }));

    //新しいエディタがインスタンス化された際のイベント処理
    this.subscriptions.add(atom.workspace.observeTextEditors(editor => {

      //もし初期化されていないViewがあれば初期化
      if(SearchStaView == null){
        SearchStaView = require('./bve-searchStaView.js');
      }
      if(this.staSearchListView == null){
        this.staSearchListView = new SearchStaView();
      }

      //文法判定
      const buffer = editor.getBuffer();
      this.setGrammar(buffer);

      //変更された際も文法を判定
      buffer.onDidChange(({changes}) => {
        this.setGrammar(buffer);
      });
    }));
  },

  deactivate() {
    this.staSearchListView.destroy();
    this.staSearchListView = null;
    this.subscriptions.dispose();
  },

  //状態を保存しておく
  serialize() {
  },

  //駅名検索パネルを表示
  openSearchStaPanel(){
    if(this.staSearchListView.isVisible){
      this.staSearchListView.hide();
    }else{
      this.staSearchListView.show(atom.workspace.getActiveTextEditor());
    };
  },

  //文法判定コマンド
  grammarJudgment() {
    const editor = atom.workspace.getActiveTextEditor();
    this.setGrammar(editor.getBuffer());
  },

  //文法判定
  setGrammar(buffer) {
    const header = buffer.lineForRow(0);
    if(header.match(/BveTs\s*Map\s*2.02/gi)){
      atom.grammars.assignLanguageMode(buffer, 'source.bve-map-2.02');
    }else if(header.match(/BveTs\s*Structure\s*List/gi)){
      atom.grammars.assignLanguageMode(buffer, 'source.bve-structure');
    }else if(header.match(/BveTs\s*Sound\s*List/gi)){
      atom.grammars.assignLanguageMode(buffer, 'source.bve-sound');
    }else if(header.match(/BveTs\s*Station\s*List/gi)){
      atom.grammars.assignLanguageMode(buffer, 'source.bve-station');
    }else if(header.match(/BveTs\s*Train/gi)){
      atom.grammars.assignLanguageMode(buffer, 'source.bve-train');
    }
  }

};


lib/bve-searchStaView.js
'use babel'

import BveSearchListView from './bve-searchListView.coffee';

export default class SearchStaView extends BveSearchListView{
  constructor() {
    super();
    super.setCallback(this.jump);
  }

  show(editor){
    var staList = this.updateStaList(editor);
    super.clearItems();
    for(var i=0; i<staList.length; i++){
      super.addItem(staList[i]);
    }
    super.show();
  }

  //駅名を配列にして取得
  updateStaList(editor){
    var text = editor.getText();
    var reg = /\s*Station\s*\[\'(.+)\'\]\.\s*Put\s*\(/gi;
    var staList = [];
    //駅名部分をキャプチャする
    while((m = reg.exec(text)) != null){
      staList.push(m[1]);
    }
    console.log(staList);
    return staList;
  }

  //駅名の行にジャンプ(コールバック)
  jump(name){
    console.log("jump to " + name);
    var editor = atom.workspace.getActiveTextEditor();
    var reg = new RegExp('\\s*Station\\s*\\\[\\\'' + name + '\\\'\\\]\\\.\\s*Put\\\(','g');
    editor.scan(reg, undefined, function(obj) {
      editor.setCursorBufferPosition(obj.range.start);
      return console.log(obj.range);
    });
  }

  // Returns an object that can be retrieved when package is activated
  serialize() {}

  // Tear down any state and detach
  destroy() {
    this.element.remove();
  }

  getElement() {
    return this.element;
  }
}


lib/bve-searchCommentView.js
'use babel'

import BveSearchListView from './bve-searchListView.coffee';
//不使用

export default class SearchCommentView extends BveSearchListView{
  constructor(serializedState) {
    super();
    super.setCallback(this.jump);
  }

  show(editor){
    var commentList = this.updateCommentList(editor);
    super.clearItems();
    for(var i=0; i<commentList.length; i++){
      super.addItem(commentList[i]);
    }
    super.show();
  }

  //駅名を配列にして取得
  updateCommentList(editor){
    var text = editor.getText();
    var reg = /(?:\/\/|#)(.+)/gi;
    var commentList = [];
    //駅名部分をキャプチャする
    while((m = reg.exec(text)) != null){
      commentList.push(m[1]);
    }
    console.log(commentList);
    return commentList;
  }

  //駅名の行にジャンプ(コールバック)
  jump(name){
    console.log("jump to " + name);
    var editor = atom.workspace.getActiveTextEditor();
    var reg = new RegExp('#' + name + '\\s*$','g');
    editor.scan(reg, undefined, function(obj) {
      editor.setCursorBufferPosition(obj.range.start);
      return console.log(obj.range);
    });
  }

  // Returns an object that can be retrieved when package is activated
  serialize() {}

  // Tear down any state and detach
  destroy() {
    this.element.remove();
  }

  getElement() {
    return this.element;
  }
}


lib/language-bve-map.js
'use babel';

import LanguageBveMapView from './language-bve-map-view';
import { CompositeDisposable } from 'atom';

export default {

  languageBveMapView: null,
  modalPanel: null,
  subscriptions: null,

  activate(state) {
    this.languageBveMapView = new LanguageBveMapView(state.languageBveMapViewState);
    this.modalPanel = atom.workspace.addModalPanel({
      item: this.languageBveMapView.getElement(),
      visible: false
    });

    atom.workspace.observeTextEditors(editor => {
      editor.onDidSave(event => {
        if(editor.lineTextForBufferRow(0).match(/BveTs\s*Map\s*2.02/)){
          editor.setGrammar(atom.grammars.grammarForScopeName('source.bve-map-2.02'));
        }else if(editor.lineTextForBufferRow(0).match(/BveTs\s*Structure\s*List\s2.00/)){
          editor.setGrammar(atom.grammars.grammarForScopeName('source.bve-structure-2.00'));
        }
      });
    });

    atom.workspace.onDidOpen(editor => {
      editor = atom.workspace.getActiveTextEditor();
      if(editor.lineTextForBufferRow(0).match(/BveTs\s*Map\s*2.02/)){
        editor.setGrammar(atom.grammars.grammarForScopeName('source.bve-map-2.02'));
      }else if(editor.lineTextForBufferRow(0).match(/BveTs\s*Structure\s*List\s2.00/)){
        editor.setGrammar(atom.grammars.grammarForScopeName('source.bve-structure-2.00'));
      }
    });

    // Events subscribed to in atom's system can be easily cleaned up with a CompositeDisposable
    this.subscriptions = new CompositeDisposable();

    // Register command that toggles this view
    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'language-bve-map:toggle': () => this.toggle()
    }));
  },

  deactivate() {
    this.modalPanel.destroy();
    this.subscriptions.dispose();
    this.languageBveMapView.destroy();
  },

  serialize() {
    return {
      languageBveMapViewState: this.languageBveMapView.serialize()
    };
  },

  toggle() {
    console.log('LanguageBveMap was toggled!');
    return (
      this.modalPanel.isVisible() ?
      this.modalPanel.hide() :
      this.modalPanel.show()
    );
  }

};
